turbopack/crates/turbopack-ecmascript/src/tree_shake/optimizations.rs

```diff
@@ -246,4 +246,205 @@ impl GraphOptimizer<'_> {
 
         did_work
     }
+
+    /// If the graph is
+    ///
+    /// a -> c -> d -> f
+    /// a -> c -> e -> f
+    /// b -> c
+    ///
+    /// all of `d, e, f` should be merged into `c` because they are all reachable only by stepping
+    /// through `c`.
+    pub(super) fn perform_dominator_analysis<N>(&self, g: &mut Graph<Vec<N>, Dependency>) -> bool
+    where
+        N: Copy,
+        Self: Index<N, Output = ItemId>,
+    {
+        let mut did_work = false;
+
+        let starting_nodes: Vec<_> = g
+            .node_indices()
+            .filter(|&node| g.edges_directed(node, Direction::Incoming).count() == 0)
+            .collect();
+
+        for &entry in &starting_nodes {
+            if g.node_weight(entry).is_none() {
+                continue;
+            }
+
+            // Skip nodes that shouldn't be merged
+            if self.should_not_merge_iter(g.node_weight(entry).expect("Node should exist")) {
+                continue;
+            }
+
+            // Create a reachability map to avoid accessing unreachable nodes
+            let mut reachable_from_entry = FxHashSet::default();
+            let mut queue = vec![entry];
+
+            while let Some(node) = queue.pop() {
+                if !reachable_from_entry.insert(node) {
+                    continue;
+                }
+
+                for edge in g.edges_directed(node, Direction::Outgoing) {
+                    queue.push(edge.target());
+                }
+            }
+
+            // Build a map of nodes that dominate other nodes
+            let mut dom_map: FxHashMap<NodeIndex, Vec<NodeIndex>> = FxHashMap::default();
+
+            // For each node, check if all paths to it go through a potential dominator
+            for node in g.node_indices() {
+                // Skip entry node and unreachable nodes
+                if node == entry || !reachable_from_entry.contains(&node) {
+                    continue;
+                }
+
+                // Find all nodes with a path to this node
+                let mut can_reach_node: FxHashSet<NodeIndex> = FxHashSet::default();
```

Comment:
Since the `NodeIndex` types are coercible to `usize`... and (afaict) they are _dense_ in the range `[0,N)` where `N` is the number of nodes in the graph.  Could we replace these hashsets with bitsets?